File: src/settings_contract.py
"""Shared settings schema helpers backed by shared/settings_contract.json."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Dict, List

_CONTRACT_PATH = Path(__file__).resolve().parents[1] / "shared" / "settings_contract.json"

_DEFAULT_CONTRACT: Dict[str, Any] = {
    "version": 1,
    "keys": [],
    "electronDefaults": {},
    "pythonDefaults": {},
    "pythonConfigKeys": [],
    "pythonConfigOptionalStringKeys": [],
}


def _load_contract() -> Dict[str, Any]:
    try:
        with _CONTRACT_PATH.open("r", encoding="utf-8") as handle:
            payload = json.load(handle)
            if isinstance(payload, dict):
                return payload
    except Exception:
        pass
    return dict(_DEFAULT_CONTRACT)


CONTRACT: Dict[str, Any] = _load_contract()
SETTINGS_SCHEMA_VERSION: int = int(CONTRACT.get("version", 1))
SETTINGS_KEYS: List[str] = list(CONTRACT.get("keys", []) or [])
ELECTRON_DEFAULTS: Dict[str, Any] = dict(CONTRACT.get("electronDefaults", {}) or {})
PYTHON_DEFAULTS: Dict[str, Any] = dict(CONTRACT.get("pythonDefaults", {}) or {})
PYTHON_CONFIG_KEYS: List[str] = list(CONTRACT.get("pythonConfigKeys", []) or [])
PYTHON_CONFIG_OPTIONAL_STRING_KEYS: List[str] = list(
    CONTRACT.get("pythonConfigOptionalStringKeys", []) or []
)

File: electron/settings_contract.js
// settings_contract.js
// Shared settings schema helpers backed by shared/settings_contract.json.

const contract = require('../shared/settings_contract.json');

const DEFAULT_CONTRACT = {
  version: 1,
  keys: [],
  electronDefaults: {},
  pythonDefaults: {},
  pythonConfigKeys: [],
  pythonConfigOptionalStringKeys: []
};

const SETTINGS_CONTRACT = Object.freeze({
  ...DEFAULT_CONTRACT,
  ...(contract || {})
});

const SETTINGS_SCHEMA_VERSION = Number.isFinite(Number(SETTINGS_CONTRACT.version))
  ? Number(SETTINGS_CONTRACT.version)
  : 1;

const SETTINGS_KEYS = Object.freeze(
  Array.isArray(SETTINGS_CONTRACT.keys) ? [...SETTINGS_CONTRACT.keys] : []
);

const ELECTRON_DEFAULTS = Object.freeze({
  ...(SETTINGS_CONTRACT.electronDefaults || {})
});

const PYTHON_DEFAULTS = Object.freeze({
  ...(SETTINGS_CONTRACT.pythonDefaults || {})
});

const PYTHON_CONFIG_KEYS = Object.freeze(
  Array.isArray(SETTINGS_CONTRACT.pythonConfigKeys)
    ? [...SETTINGS_CONTRACT.pythonConfigKeys]
    : []
);

const PYTHON_CONFIG_OPTIONAL_STRING_KEYS = new Set(
  Array.isArray(SETTINGS_CONTRACT.pythonConfigOptionalStringKeys)
    ? SETTINGS_CONTRACT.pythonConfigOptionalStringKeys
    : []
);

module.exports = {
  SETTINGS_CONTRACT,
  SETTINGS_SCHEMA_VERSION,
  SETTINGS_KEYS,
  ELECTRON_DEFAULTS,
  PYTHON_DEFAULTS,
  PYTHON_CONFIG_KEYS,
  PYTHON_CONFIG_OPTIONAL_STRING_KEYS
};
